crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -1419,6 +1420,360 @@ impl<'w, 'a, T: Component> VacantEntry<'w, 'a, T> {
+            return None;
+        };
+        self.access
+            .has_read(id)
+            // SAFETY: We have read access so we must have the component
+            .then(|| unsafe { self.entity.get_change_ticks::<T>().debug_checked_unwrap() })
+    }
+
+    /// Retrieves the change ticks for the given [`ComponentId`]. This can be useful for implementing change
+    /// detection in custom runtimes.
+    ///
+    /// **You should prefer to use the typed API [`Self::get_change_ticks`] where possible and only
+    /// use this in cases where the actual component types are not known at
+    /// compile time.**
+    #[inline]
+    pub fn get_change_ticks_by_id(&self, component_id: ComponentId) -> Option<ComponentTicks> {
+        // SAFETY: We have read access so we must have the component
+        self.access.has_read(component_id).then(|| unsafe {
+            self.entity
+                .get_change_ticks_by_id(component_id)
+                .debug_checked_unwrap()
+        })
+    }
+
+    /// Gets the component of the given [`ComponentId`] from the entity.
+    ///
+    /// **You should prefer to use the typed API [`Self::get`] where possible and only
+    /// use this in cases where the actual component types are not known at
+    /// compile time.**
+    ///
+    /// Unlike [`FilteredEntityRef::get`], this returns a raw pointer to the component,
+    /// which is only valid while the [`FilteredEntityRef`] is alive.
+    #[inline]
+    pub fn get_by_id(&self, component_id: ComponentId) -> Option<Ptr<'w>> {
+        self.access
+            .has_read(component_id)
+            // SAFETY: We have read access so we must have the component
+            .then(|| unsafe { self.entity.get_by_id(component_id).debug_checked_unwrap() })
+    }
+}
+
+impl<'w> From<FilteredEntityMut<'w>> for FilteredEntityRef<'w> {
+    fn from(entity_mut: FilteredEntityMut<'w>) -> Self {
+        // SAFETY:
+        // - `EntityWorldMut` guarantees exclusive access to the entire world.
+        unsafe { FilteredEntityRef::new(entity_mut.entity, entity_mut.access) }
+    }
+}
+
+impl<'a> From<&'a FilteredEntityMut<'_>> for FilteredEntityRef<'a> {
+    fn from(entity_mut: &'a FilteredEntityMut<'_>) -> Self {
+        // SAFETY:
+        // - `EntityWorldMut` guarantees exclusive access to the entire world.
+        unsafe { FilteredEntityRef::new(entity_mut.entity, entity_mut.access.clone()) }
+    }
+}
+
+/// Provides mutable access to a single entity and some of its components defined by the contained [`Access`].
+pub struct FilteredEntityMut<'w> {
+    entity: UnsafeEntityCell<'w>,
+    access: Access<ComponentId>,
+}
+
+impl<'w> FilteredEntityMut<'w> {
+    /// # Safety
+    /// - `cell` must have permission to mutate every component of the entity.
+    /// - No accesses to any of the entity's components may exist
+    ///   at the same time as the returned [`FilteredEntityMut`].
```

Comment:
```suggestion
    /// - No `&mut World` can exist from the underlying `UnsafeWorldCell`
    /// - If `acccess` takes read access to a component no mutable reference to that 
    /// component can exist at the same time as the returned [`FilteredEntityMut`]
    /// - If `access` takes write access to a component, no reference to that component
    /// may exist at the same time as the returned [`FilteredEntityMut`]
```
Is this more what the safety requirements are here?